contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+Trusted clients should be able to modify the initialization and external admission hooks on the fly and expect
+that configuration is updated quickly. Extension API servers should also be able to leverage central
+configuration, but may opt for alternate mechanisms.
+
+The initializer admission controller and the generic webhook admission controller should dynamically load config
+from a `ConfigMap` or a net new API object holding the following configuration schema:
+
+```
+type AdmissionControlConfiguration struct {
+    TypeMeta // although this object could simply be serialized like ComponentConfig
+
+    // ResourceInitializers is a list of resources and their default initializers
+    ResourceInitializers []ResourceDefaultInitializer
+
+    ExternalAdmissionHooks []ExternalAdmissionHook
+}
+
+type ResourceDefaultInitializer struct {
+    // Resource identifies the type of resource to be initialized that should be initialized
+    Resource GroupResource
+    // Initializers are the default names that will be registered to this resource
+    Initializers []string
+}
+
+type ExternalAdmissionHook struct {
+    // Operations is the list of operations this hook will be invoked on - Create, Update, or *
+    // for all operations. Defaults to '*'.
+    Operations []string
+    // Resources are the resources this hook should be invoked on. '*' is all resources.
+    Resources []string
+    // Subresources are the list of subresources this hook should be invoked on. '*' is all resources.
+    Subresources []string
+
+    // TODO define client configuration
+
+    // FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+    // allowed values are Ignore, Retry, Fail. Default value is Fail
+    FailurePolicy FailurePolicyType
+}
+```
+
+All changes to this config must be done in an safe matter - when adding a new hook or initializer,
+first verify the new agent is online before allowing it to come into rotation. Removing a hook or
+initializer must occur before disabling the remote endpoint, and all queued items must complete.
+
+
+### Alternatives considered
+
+The following are all viable alternatives to this specification, but have some downsides against the requirements above.
+There should be no reason these could not be implemented for specific use cases.
+
+1.  Admission controller that can run shell commands inside its context to mutate objects.
+    * Limits on performance and reliability
+    * Requires the masters be updated (can't be done dynamically)
+2. Admission controller that can run a scripting language like Lua or JavaScript in process.
+    * Limits on performance and reliability
+    * Not consistent with existing tools and infrastructure
+    * Requires that masters be updated and has limits on dynamic behavior
+3. Direct external call outs for object mutation (RPC to initialize objects)
+    * Requires a new programming model
+    * Duplicates our create - watch - update logic from controllers
```

Comment:
We already have lots of objects that have quasi initialization behavior - pods before they are scheduled, services before load balancers are allocated, routes before a router sees them.  We have to formalize some chain of custody and extension of phase, and I'd rather have one approach (initializers) rather than one per resource.

* Services need IPs filled out - has to be before clients observe the result
* Vertical autosizing has to happen before scheduling

I think RPC has advantages in some cases.  I'm not as far on its side though as our existing patterns.